### PR TITLE
Many changes and some fixes

### PR DESCRIPTION
**Pull request by joevt**
All the changes are listed in the Changelog.md. The following is some of the more important ones:
- Use edk2 for path conversion (supports all known node types).
- Added the ability to convert a single efi device path from bin, hex, or nvram format to string or from string to bin. The input can be an argument or stdin. stdin is required for binary path.
- Fixed the format of ioregistry output so that the ioregistry path is complete and can be used to convert an ioregistry device to an efi path.
- efi path node sizes are now verified.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -1,5 +1,38 @@
 gfxutil changelog
 =================
+#### 1.79b joevt Dec 24, 2019
```
Comment (vit9696) [inline Changelog.md]: Could you make it 1.80b without suffixes as we now mainline the changes.

File: edk2misc.c [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+//  gfxutil
+//
+//  Created by joevt on 2019-12-14.
+//
+
+#include "edk2misc.h"
+
+//#include <stdio.h>
+
+// #undef __pic__
```
Comment (vit9696) [inline edk2misc.c]: Please remove commented out code here, above and below.

File: edk2misc.c [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+VOID *
+InternalAllocateCopyPool (
+   UINTN            AllocationSize,
+   CONST VOID       *Buffer
+  )
+{
+  VOID  *Memory;
+
+  ASSERT (Buffer != NULL);
+//  ASSERT (AllocationSize <= (MAX_ADDRESS - (UINTN) Buffer + 1));
```
Comment (vit9696) [inline edk2misc.c]: Hrm, maybe remove that?

File: edk2misc.c [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+
+EFI_STATUS
+EFIAPI
+DontHandleProtocol (
+  IN EFI_HANDLE       UserHandle,
+  IN EFI_GUID         *Protocol,
+  OUT VOID            **Interface
+  )
+{
+	return EFI_UNSUPPORTED;
```
Comment (vit9696) [inline edk2misc.c:80]: Could you please fix tab/space consistency?

**Comment by joevt** [comment]
One thing I would have liked to have done is to find a better way to modify mUefiDevicePathLibToTextTable (which is const) without modifying any files in edk2.

What I did instead was created a patched version of ConvertDevicePathToText which uses a copied version of mUefiDevicePathLibToTextTable.

File: edk2misc.c [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+
+typedef struct {
+	UINT8                Type;
+	UINT8                SubType;
+	int                  SizeMin;
+	int                  SizeMax; // 0 = same as min, 1 = infinite
+	int                  SizeInc;
+} DEVICE_NODE_TO_SIZE;
+
+//const int x = sizeof(DEVICE_NODE_TO_SIZE.SizeMin);
```
Comment (vit9696) [inline edk2misc.c]: This looks like debug legacy.

File: edk2misc.c [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+	//
+	gST = &ST;
+
+	//
+	// Cache pointer to the EFI Boot Services Table
+	//
+	gBS = gST->BootServices;
+	
+	/*
+	printf("hello4: mmap(%llx) %llx : %d\n",
```
Comment (vit9696) [inline edk2misc.c]: This looks like debug legacy.

File: main.h [hunk 7]
```diff
@@ -7,13 +7,15 @@
...
  */
 
+#ifndef _MAIN_H
+#define _MAIN_H
+
+#include "UefiDevicePathLib.h"
+
 // Constants
 #define MAX_FILENAME 255
 #define VERSION "1.79b"
```
Comment (vit9696) [inline main.h]: And update this version to 1.80b

**Review by vit9696 — COMMENTED** [review]
Hi, I marked some notes, but it is better than before anyway. Where are the EDK II files? I did not notice them in the diff…

**Comment by joevt** [comment]
edk2 is from https://github.com/tianocore/edk2

I have gfxutil folder beside the edk2 folder like how Clover is done (except Clover is placed inside the edk2 folder or has it changed since the last time I tried to compile it?).

This was done so that edk2 files would not need to be copied into gfxutil.

**Comment by vit9696** [comment]
Right, so we can clone EDK II during the build process. Ok, this makes good sense to me.

> One thing I would have liked to have done is to find a better way to modify mUefiDevicePathLibToTextTable (which is const) without modifying any files in edk2.

> What I did instead was created a patched version of ConvertDevicePathToText which uses a copied version of mUefiDevicePathLibToTextTable.

I guess this is quite ugly, but I do not think I can imagine better ways to approach this. Could define CONST as empty, but I believe it can break other places too. Probably it is the most reasonable at the moment.

**Comment by joevt** [comment]
> I guess this is quite ugly, but I do not think I can imagine better ways to approach this. Could define CONST as empty, but I believe it can break other places too. Probably it is the most reasonable at the moment.

I was wondering if edk2 could be built as library and if there was a linker option to export a symbol as not const but I couldn't find a method to do that or I didn't look very hard because it was not obvious enough.

We can't declare it not const elsewhere because linker places const stuff in read only segment so modifications would cause an exception/signal.

**Comment by joevt** [comment]
One thing I didn't get done was modifying the man page because I don't know the format. I don't think there's a build instruction to install the man page? The help option -h has all the other options.

File: Changelog.md (hunk 1 above)
Comment (joevt) [reply Changelog.md]: Fixed.

File: edk2misc.c (hunk 2 above)
Comment (joevt) [reply edk2misc.c]: Fixed.

File: edk2misc.c (hunk 3 above)
Comment (joevt) [reply edk2misc.c]: Fixed.

File: edk2misc.c (hunk 4 above)
Comment (joevt) [reply edk2misc.c:80]: Fixed.

File: edk2misc.c (hunk 5 above)
Comment (joevt) [reply edk2misc.c]: Fixed.

File: edk2misc.c (hunk 6 above)
Comment (joevt) [reply edk2misc.c]: Fixed.

File: main.h (hunk 7 above)
Comment (joevt) [reply main.h]: Fixed.

**Comment by vit9696** [comment]
> One thing I didn't get done was modifying the man page because I don't know the format. I don't think there's a build instruction to install the man page? The help option -h has all the other options

This can be lived with, as I do not quite have time to invest into updating the man page myself.

The bigger issue is sanitizers and analysers. I fixed some of the bugs spotted in our code, but there are several others that are present in EDK II.

For example, UBSan moans about incorrect unaligned implementation:

```
edk2/MdePkg/Library/BaseLib/Unaligned.c:188:10: runtime error: load of misaligned address 0x604000069294 for type 'const UINT64' (aka 'const unsigned long long'), which requires 8 byte alignment
0x604000069294: note: pointer points here
  03 0a 2c 00 1b 61 99 42  c9 4c df 43 88 2b 2a 07  07 a4 a9 5f 53 da cd 73  f2 95 b4 42 bc 87 6e 5f
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior edk2/MdePkg/Library/BaseLib/Unaligned.c:188:10 in 
```

And Clang Static Analyzer finds dead code due to Pcd definition and failure to see `Flags |= ARGUMENT_REVERSED;`:

```
edk2/MdePkg/Library/UefiDevicePathLib/DevicePathUtilities.c:68:8: Value stored to 'Count' is never read
edk2/MdePkg/Library/BasePrintLib/PrintLibInternal.c:1107:68: The left operand of '!=' is a garbage value
edk2/MdePkg/Library/BasePrintLib/PrintLibInternal.c:1110:148: The left operand of '<<' is a garbage value
```

I believe we should somehow add this to exceptions or avoid the analysis, as I do not want to entirely kill Clang Static Analyzer. Any ideas?

**Comment by roddy20** [comment]
[build errors.zip](https://github.com/acidanthera/gfxutil/files/4094183/build.errors.zip)
** BUILD FAILED **

**Comment by vit9696** [comment]
Builds fine, you have to clone edk2 before building. https://github.com/acidanthera/gfxutil/blob/master/.travis.yml#L11

**Comment by roddy20** [comment]
** BUILD SUCCEEDED **
works OK
/Users/rodion/src/gfxutil/build/Release/gfxutil -p
00:00.0 8086:29e0 /PCI0/MCHC@0 = PciRoot(0x0)/Pci(0x0,0x0)
00:01.0 8086:29e1 /PCI0/P0P2@1 = PciRoot(0x0)/Pci(0x1,0x0)
00:06.0 8086:29e9 /PCI0/P0P3@6 = PciRoot(0x0)/Pci(0x6,0x0)
00:1a.0 8086:3a37 /PCI0/UHC4@1A = PciRoot(0x0)/Pci(0x1A,0x0)
00:1a.2 8086:3a39 /PCI0/UHC6@1A,2 = PciRoot(0x0)/Pci(0x1A,0x2)
00:1a.1 8086:3a38 /PCI0/UHC5@1A,1 = PciRoot(0x0)/Pci(0x1A,0x1)
00:1a.7 8086:293c /PCI0/EHC2@1A,7 = PciRoot(0x0)/Pci(0x1A,0x7)
00:1b.0 8086:293e /PCI0/HDEF@1B = PciRoot(0x0)/Pci(0x1B,0x0)
etc

**Comment by vit9696** [comment]
@joevt, actually, we spotted two issues after merging.

1. By default the tool now does device path conversion, and it may be a little undesired to change the original behaviour that printed device path report like with `-p` key. Do you really need that logic, I get a feeling it can be done with a separate argument, as it is needed less frequently.

2. AddressSanitizer (enabled in diagnostics) finds memory corruptions, caused by most likely missing calls to `IsDevicePathValid` and `\0` terminations:

```
$ echo 123 | ./gfxutil
=================================================================
==50280==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000004ba at pc 0x0001078cabd7 bp 0x7ffee83a7fa0 sp 0x7ffee83a7f98
WRITE of size 2 at 0x6020000004ba thread T0
    #0 0x1078cabd6 in AsciiStrToUnicodeStr String.c:1743
    #1 0x1078e98fc in OutputDevicePathUtilFromText efidevp.c:83
    #2 0x1078eba64 in parse_generic_option efidevp.c:186
    #3 0x10787e9dd in parse_args main.c:1313
    #4 0x107885583 in main main.c:1737
    #5 0x1078569d3 in start (gfxutil:x86_64+0x1000019d3)

0x6020000004ba is located 0 bytes to the right of 10-byte region [0x6020000004b0,0x6020000004ba)
allocated by thread T0 here:
    #0 0x108799793 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x61793)
    #1 0x1078abbf4 in AllocatePool MemoryAllocationLib.c:33
    #2 0x1078e98e0 in OutputDevicePathUtilFromText efidevp.c:80
    #3 0x1078eba64 in parse_generic_option efidevp.c:186
    #4 0x10787e9dd in parse_args main.c:1313
    #5 0x107885583 in main main.c:1737
    #6 0x1078569d3 in start (gfxutil:x86_64+0x1000019d3)

SUMMARY: AddressSanitizer: heap-buffer-overflow String.c:1743 in AsciiStrToUnicodeStr
Shadow bytes around the buggy address:
  0x1c0400000040: fa fa fd fd fa fa fd fa fa fa fd fd fa fa fd fd
  0x1c0400000050: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fd
  0x1c0400000060: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fa
  0x1c0400000070: fa fa fd fd fa fa fd fa fa fa fd fd fa fa fd fa
  0x1c0400000080: fa fa fd fd fa fa 00 fa fa fa 01 fa fa fa fd fa
=>0x1c0400000090: fa fa fd fd fa fa 00[02]fa fa fa fa fa fa fa fa
  0x1c04000000a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c04000000b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c04000000c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c04000000d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c04000000e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==50280==ABORTING
Abort trap: 6
```

Could you submit another PR fixing it?

Actually I resolved the memory corruption issues I spotted in https://github.com/acidanthera/gfxutil/commit/00c270fe86e326462e666a73c3e746d28b900741, but the other question still holds.

**Comment by joevt** [comment]
I like the -p option to dump all paths (like nvram -p dumps all nvram parameters). I think that no arguments should do nothing except output the usage? That seems the safest way. Or we can go back to the original behavior where no arguments outputs all paths.

For path conversion, we can add a new argument, maybe "-c". Path conversion allows the command to work with paths that don't exist on your computer (from someone else's ioreg or nvram output for example) or to convert paths that are stored elsewhere (inside Boot0000, etc.)

**Comment by vit9696** [comment]
Running without an arguments to print all paths was added to let any people (potentially less familiar with Terminal) to see their devices by merely double-clicking the tool. While it may seem unreasonable, this is very likely the most common action performed at least compared to conversion, which is somewhat specific.

All in all, I like the idea that `echo Something | gfxutil` works as a pipe, but I do not like running `gfxutil` without arguments waiting for stdin. This does not look much Unix way.

I suggest to:
- make path conversion work via `-c`
- leave support for piping without arguments by detecting Terminal presence via `isatty(fileno(stdin))`
- make `-p` the default option

**Comment by joevt** [comment]
I tried ```isatty(fileno(stdin))``` but it wouldn't work correctly from a BBEdit worksheet. I'll try again. The workaround would be to always use an argument if you like BBEdit worksheets: isatty won't be used/required when certain action arguments exist.
I'll implement these changes.

**Comment by vit9696** [comment]
Quite possible that for some IDEs it will not work as they may not emulate terminal. I guess, it makes sense not to remove -p option for such cases, so that when running with no arguments does not work, one can still force output by adding -p. The rest holds though. Thanks.

**Comment by vit9696** [comment]
@joevt do you have an estimate for the patch? We plan a release early in February, and definitely want to include it. Thanks.

**Comment by joevt** [comment]
I can work on it tomorrow.

**Comment by vit9696** [comment]
Ah, that is more than fine. Thanks.

**Pull request merged**